### PR TITLE
Capture RNG resource via batched driver constructor

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
A step closer to clean up RNG handling in drivers.
1. Capture RNG resource via batched driver constructor. No longer directly access `RandomNumberControl::Children`.
2. Add a RandomNumberGeneratorPool helper class for unit test need of fake RNGs.
3. There are small changes moving step_contexts_ to VMC and DMC drivers that are unrelated to this topic but I was unable to take them out of this PR.

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
laptop

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted

### CODE REVIEW

**Comment by prckent** [comment]
Still draft?

**Comment by ye-luo** [comment]
> Still draft?

It is ready for review.

File: src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h [hunk 1]
```diff
@@ -57,6 +57,7 @@ class QMCFixedSampleLinearOptimizeBatched : public QMCDriverNew, LinearMethod
                                       VMCDriverInput&& vmcdriver_input,
                                       WalkerConfigurations& wc,
                                       MCPopulation&& population,
+             const RefVector<RandomBase<FullPrecRealType>>& rng_refs,
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCFixedSampleLinearOptimizeBatched.h:60]: will fix the formatting in my next PR.

**Comment by ye-luo** [comment]
Test this please

**Review by prckent — APPROVED** [review]

**Pull request merged**